### PR TITLE
Multiple blocks mirror simultaneously

### PR DESCRIPTION
**Pull request by zhencliu**
ID: 1875215

### CODE REVIEW

**Comment by zhencliu** [comment]
local fs:
 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (402.78 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-06T05.47-b4174c7/results.html
JOB TIME   : 408.70 s

iscsi direct:
 (1/1) Host_RHEL.m8.u3.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (396.96 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-06T06.14-d0d68e8/results.html
JOB TIME   : 402.83 s

nbd:
 (1/1) Host_RHEL.m8.u3.product_av.nbd.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (544.84 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-06T06.35-edea690/results.html
JOB TIME   : 550.65 s

gluster direct:
 (1/1) Host_RHEL.m8.u2.product_av.gluster_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (566.56 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2020-09-06T18.55-a371233/results.html
JOB TIME   : 571.71 s

hi @aliang123 @YongxueHong , can you please help to review the patch? Thanks much.

File: provider/blockdev_mirror_base.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+import six
```
[the remark this answers is not in the document]
Comment (zhencliu) [reply provider/blockdev_mirror_base.py:1]: done

File: provider/blockdev_mirror_base.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    def blockdev_mirror(self):
+        """Need to be implemented in specific test case"""
+        raise NotImplementedError
+
+    def do_test(self):
+        self.blockdev_mirror()
+        self.check_mirrored_block_nodes_attached()
+        self.clone_vm_with_mirrored_images()
+        self.verify_data_files()
+        self.remove_files_from_system_image()
```
[the remark this answers is not in the document]
Comment (zhencliu) [reply provider/blockdev_mirror_base.py:192]: Hi Yongxue, do_test was implemented here because most block-mirror test cases follow the same test steps, just like what do_test does here, for others, they almost follow the same steps, with just a little difference, e.g. create a new file after block-mirror, so if we implement do_test in specific test case, we have to duplicate many code lines...

The difference between these test cases is the block_mirror, I have implemented three bock_mirror_* functions, which can be called in specific test cases.

This class likes a virtual c++ abstract class, it cannot be used to define an BlockdevMirrorBaseTest object, it can only be derived by children --- the specific test cases.

File: qemu/tests/blockdev_mirror_multiple_blocks.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        session = self.main_vm.wait_for_login()
+        try:
+            disk_params = self.params.object_params(tag)
+            disk_size = disk_params["image_size"]
+            disks = utils_disk.get_linux_disks(session, True)
+
+            for kname, attr in six.iteritems(disks):
+                if attr[1] == disk_size and attr[2] == "disk":
+                    # skip disks in use
+                    disk_path = "/dev/%s1" % kname
```
[the remark this answers is not in the document]
Comment (zhencliu) [reply qemu/tests/blockdev_mirror_multiple_blocks.py:20]: sda will be filtered by disk size, here skip data disks in use, e.g.  two data disks with a 2G size, sdb and sdc, we cannot differentiate them by disk size, so I added this check.

In fact /dev/sdxx will be mounted in another function generate_data_files, but it's bad to make the deps.

Anyway, updated code with get_linux_disks(session, False), which list only disks with no partition.

[the remark this answers is not in the document]
Comment (zhencliu) [reply qemu/tests/blockdev_mirror_multiple_blocks.py:20]: I am afraid I cannot do that (get_linux_disks(session, False)).
For nbd backends, the exported image cannot be created by qemu-img, kvm employs the images directly, so if the image was used before and partitioned, get_linux_disks(session, False) will return empty dict

File: provider/blockdev_mirror_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (YongxueHong) [reply provider/blockdev_mirror_base.py:205]: Okay, I got your idea, looks good to me.

File: qemu/tests/blockdev_mirror_multiple_blocks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhencliu) [reply qemu/tests/blockdev_mirror_multiple_blocks.py]: Talked with Yongxue, we made a decision that marking the disks with serial/wwn, so we can retrieve the disk with serial/wwn. More readable and simple.

**Comment by zhencliu** [comment]
(1/1) Host_RHEL.m8.u3.product_av.nbd.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (430.17 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-10T04.45-63b3f41/results.html
JOB TIME   : 435.97 s

**Comment by zhencliu** [comment]
hi @YongxueHong , can you please help double check , thanks.

**Review by YongxueHong — APPROVED** [review]

**Comment by zhencliu** [comment]
hi @YongxueHong , refer to @aliang123 's comments, I removed the three blockdev_mirror_* functions from blockdev_mirror_base, and created three new modules which derived from blockdev_mirror_base: blockdev_mirror_nowait/blockdev_mirror_wait/blockdev_mirror_parallel, which are specified for the following three mirror tests:
1. run mirror in parallel
2. run mirror and wait mirror done
3. run mirror without waiting mirror done
These three modules can cover almost all mirror test cases, for the left corner cases, I will create test case derived from blockdev_mirror_base directly.

The advantage: 1) removing dup code from test cases 2) the mirror operation can be explicitly specified by class name instead of blockdev_mirror function, e.g. class BlockdevMirrorRebootTest(BlockdevMirrorParallelTest)
The disadvantage: 1) add a mid-level inheritance class between test case and blockdev_mirror_base 2) for corner cases, we still need to derive from blockdev_mirror_base

It's OK for me to make the change, can you please review the patch again? Thanks.

**Comment by YongxueHong** [comment]
> hi @YongxueHong , refer to @aliang123 's comments, I removed the three blockdev_mirror_* functions from blockdev_mirror_base, and created three new modules which derived from blockdev_mirror_base: blockdev_mirror_nowait/blockdev_mirror_wait/blockdev_mirror_parallel, which are specified for the following three mirror tests:
> 
>     1. run mirror in parallel
> 
>     2. run mirror and wait mirror done
> 
>     3. run mirror without waiting mirror done
>        These three modules can cover almost all mirror test cases, for the left corner cases, I will create test case derived from blockdev_mirror_base directly.
> 
> 
> The advantage: 1) removing dup code from test cases 2) the mirror operation can be explicitly specified by class name instead of blockdev_mirror function, e.g. class BlockdevMirrorRebootTest(BlockdevMirrorParallelTest)
> The disadvantage: 1) add a mid-level inheritance class between test case and blockdev_mirror_base 2) for corner cases, we still need to derive from blockdev_mirror_base
> 
> It's OK for me to make the change, can you please review the patch again? Thanks.
> commit: [1bc8ccc](https://github.com/autotest/tp-qemu/commit/1bc8cccae315fd6ef12575b33e8ea51e2e373556)

Looks good to me.
Acked.

**Comment by zhencliu** [comment]
hi @ngu-niny , can you please help to review this patch against ppc? Thanks much.

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks.q35: PASS (604.31 s)

Acked-by: Aihua Liang<aliang@redhat.com>

**Comment by ngu-niny** [comment]
LGTM and test PASS on ppc:

 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_multiple_blocks: PASS (564.09 s)

Acked-by: Nini Gu <ngu@redhat.com>

**Comment by zhencliu** [comment]
Thank you all. Hi @YongxueHong , all acked, can you please help to merge it? Thanks.
